scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala

```diff
@@ -329,4 +329,283 @@ object OptimizationRules {
           }
         })
     }
+
+  /////////////////////////////
+  //
+  // Here are some actual rules for simplifying TypedPipes
+  //
+  /////////////////////////////
+
+  /**
+   * a.flatMap(f).flatMap(g) == a.flatMap { x => f(x).flatMap(g) }
+   */
+  object ComposeFlatMap extends PartialRule[TypedPipe] {
+    def applyWhere[T](on: Dag[TypedPipe]) = {
+      case FlatMapped(FlatMapped(in, fn0), fn1) =>
+        FlatMapped(in, FlatMappedFn(fn1).runAfter(FlatMapping.FlatM(fn0)))
+      case FlatMapValues(FlatMapValues(in, fn0), fn1) =>
+        FlatMapValues(in, FlatMappedFn(fn1).runAfter(FlatMapping.FlatM(fn0)))
+    }
+  }
+
+  /**
+   * a.map(f).map(g) == a.map { x => f(x).map(g) }
+   */
+  object ComposeMap extends PartialRule[TypedPipe] {
+    def applyWhere[T](on: Dag[TypedPipe]) = {
+      case Mapped(Mapped(in, fn0), fn1) =>
+        Mapped(in, ComposedMapFn(fn0, fn1))
+      case MapValues(MapValues(in, fn0), fn1) =>
+        MapValues(in, ComposedMapFn(fn0, fn1))
+    }
+  }
+
+  /**
+   * a.filter(f).filter(g) == a.filter { x => f(x) && g(x) }
+   */
+  object ComposeFilter extends Rule[TypedPipe] {
+    def apply[T](on: Dag[TypedPipe]) = {
+      // scala can't type check this, so we hold its hand:
+      // case Filter(Filter(in, fn0), fn1) =>
+      //   Some(Filter(in, ComposedFilterFn(fn0, fn1)))
+      case f@Filter(_, _) =>
+        def go[A](f: Filter[A]): Option[TypedPipe[A]] =
+          f.input match {
+            case f1: Filter[a] =>
+              Some(Filter[a](f1.input, ComposedFilterFn(f.fn, f.fn)))
+            case _ => None
+          }
+        go(f)
+      case FilterKeys(FilterKeys(in, fn0), fn1) =>
+        Some(FilterKeys(in, ComposedFilterFn(fn0, fn1)))
+      case _ => None
+    }
+  }
+
+  /**
+   * a.onComplete(f).onComplete(g) == a.onComplete { () => f(); g() }
```

Comment:
when/where do we use an OnComplete (i.e. given its a side effect do we need to ensure all of these are ran // do try/error handling around each invocation?)

Reply:
I don't think we ever do. Users at Twitter used this to make some call to an API after they finished doing some side-effecting map or reduce. Other users have asked for it. I kind of hate the API now since I think it is an anti-pattern, but alas, I don't want to remove it (and we can optimize it, since we can move it to the very end of a map phase commuting with other map operations.